File: src/domain/motes.ts
/**
 * Motes: bindable elemental spirits.
 *
 * A mote sits in one of three states.
 *
 *   set        - bound to an actor. Contributes its stat bonus and counts
 *                toward the actor's class.
 *   standby    - it has been unleashed this battle. Its bonus and class
 *                contribution are gone until it returns, but it can be spent
 *                to call a summon.
 *   recovering - spent on a summon. Returns to `set` after a few rounds.
 *
 * The loop is the interesting part: unleashing a mote is an immediate tactical
 * gain that costs you stats and can knock you out of your class mid-fight, and
 * summoning cashes those in for a large hit at the cost of a longer downtime.
 * Every state change here is reversible and fully serializable.
 */

import type { Element } from './elements.js';
import type { StatModifier } from './stats.js';

export type MoteState = 'set' | 'standby' | 'recovering';

/** What happens when a mote is unleashed in battle. */
export type MoteEffect =
  | { kind: 'damage'; power: number; targeting: MoteTargeting; ignoresDefense?: boolean }
  | { kind: 'heal'; power: number; targeting: MoteTargeting }
  | { kind: 'revive'; hpFraction: number }
  | { kind: 'status'; statusId: string; chance: number; targeting: MoteTargeting }
  | { kind: 'buff'; modifier: StatModifier; rounds: number; targeting: MoteTargeting }
  | { kind: 'restore-aether'; amount: number; targeting: MoteTargeting }
  | { kind: 'utility'; tag: string };

export type MoteTargeting = 'one-foe' | 'all-foes' | 'self' | 'one-ally' | 'all-allies';

export interface MoteDef {
  id: string;
  name: string;
  element: Element;
  description?: string;
  /** Applied to the bearer while the mote is `set`. */
  bonus?: StatModifier;
  /** Resolved when the mote is unleashed. */
  effect: MoteEffect;
  /** Rounds spent `recovering` after being spent on a summon. */
  recoveryRounds?: number;
  /** Optional field ability granted while set, e.g. for overworld puzzles. */
  fieldAbilityId?: string;
}

export interface MoteInstance {
  defId: string;
  state: MoteState;
  /** Rounds left in `recovering`. Zero in every other state. */
  recoveryLeft: number;
}

export const DEFAULT_RECOVERY_ROUNDS = 3;

export function makeMote(defId: string, state: MoteState = 'set'): MoteInstance {
  return { defId, state, recoveryLeft: 0 };
}

export function isSet(mote: MoteInstance): boolean {
  return mote.state === 'set';
}

export function isStandby(mote: MoteInstance): boolean {
  return mote.state === 'standby';
}

/** Motes currently contributing to class and stats, grouped by element. */
export function setCounts(
  motes: readonly MoteInstance[],
  defOf: (id: string) => MoteDef | undefined,
): Record<Element, number> {
  const counts = { terra: 0, pyre: 0, aeris: 0, rime: 0 } as Record<Element, number>;
  for (const mote of motes) {
    if (mote.state !== 'set') continue;
    const def = defOf(mote.defId);
    if (def) counts[def.element] += 1;
  }
  return counts;
}

/** Motes on standby, which are the currency for summons. */
export function standbyCounts(
  motes: readonly MoteInstance[],
  defOf: (id: string) => MoteDef | undefined,
): Record<Element, number> {
  const counts = { terra: 0, pyre: 0, aeris: 0, rime: 0 } as Record<Element, number>;
  for (const mote of motes) {
    if (mote.state !== 'standby') continue;
    const def = defOf(mote.defId);
    if (def) counts[def.element] += 1;
  }
  return counts;
}

/** Move a set mote to standby. Returns false if it was not available. */
export function unleashMote(motes: MoteInstance[], defId: string): boolean {
  const mote = motes.find((m) => m.defId === defId && m.state === 'set');
  if (!mote) return false;
  mote.state = 'standby';
  mote.recoveryLeft = 0;
  return true;
}

/**
 * Spend standby motes of a given element on a summon, moving them to
 * `recovering`. Returns the ids actually spent, which is empty if the actor
 * could not cover the cost.
 */
export function spendStandby(
  motes: MoteInstance[],
  element: Element,
  count: number,
  defOf: (id: string) => MoteDef | undefined,
): string[] {
  const available = motes.filter((m) => m.state === 'standby' && defOf(m.defId)?.element === element);
  if (available.length < count) return [];
  const spent: string[] = [];
  for (let i = 0; i < count; i++) {
    const mote = available[i] as MoteInstance;
    mote.state = 'recovering';
    mote.recoveryLeft = defOf(mote.defId)?.recoveryRounds ?? DEFAULT_RECOVERY_ROUNDS;
    spent.push(mote.defId);
  }
  return spent;
}

/** End-of-round recovery tick. Returns ids that returned to `set`. */
export function tickRecovery(motes: MoteInstance[]): string[] {
  const recovered: string[] = [];
  for (const mote of motes) {
    if (mote.state !== 'recovering') continue;
    mote.recoveryLeft -= 1;
    if (mote.recoveryLeft <= 0) {
      mote.state = 'set';
      mote.recoveryLeft = 0;
      recovered.push(mote.defId);
    }
  }
  return recovered;
}

/** Between battles every mote returns to its bearer. */
export function restoreAllMotes(motes: MoteInstance[]): void {
  for (const mote of motes) {
    mote.state = 'set';
    mote.recoveryLeft = 0;
  }
}

File: unity/Aetherlight.Core/Domain/Motes.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Aetherlight.Domain
{
    /// <summary>
    /// A mote sits in one of three states.
    ///
    /// Set        - bound to an actor. Contributes its stat bonus and counts
    ///              toward the actor's class.
    /// Standby    - unleashed this battle. Its bonus and class contribution are
    ///              gone until it returns, but it can be spent on a summon.
    /// Recovering - spent on a summon. Returns to Set after a few rounds.
    /// </summary>
    public enum MoteState
    {
        Set,
        Standby,
        Recovering,
    }

    public sealed class MoteDef
    {
        public string Id = "";
        public string Name = "";
        public Element Element;
        public StatModifier? Bonus;
        public int RecoveryRounds = Motes.DefaultRecoveryRounds;
        public string? FieldAbilityId;
    }

    public sealed class MoteInstance
    {
        public string DefId = "";
        public MoteState State = MoteState.Set;
        /// <summary>Rounds left in Recovering. Zero in every other state.</summary>
        public int RecoveryLeft;

        public MoteInstance() { }

        public MoteInstance(string defId, MoteState state = MoteState.Set)
        {
            DefId = defId;
            State = state;
        }

        public MoteInstance Clone() => new MoteInstance { DefId = DefId, State = State, RecoveryLeft = RecoveryLeft };
    }

    public static class Motes
    {
        public const int DefaultRecoveryRounds = 3;

        /// <summary>Motes currently contributing to class and stats, by element.</summary>
        public static ElementTable<int> SetCounts(IEnumerable<MoteInstance> motes, Func<string, MoteDef?> defOf)
        {
            var counts = ElementTable<int>.Filled(0);
            foreach (var mote in motes)
            {
                if (mote.State != MoteState.Set) continue;
                var def = defOf(mote.DefId);
                if (def != null) counts[def.Element] += 1;
            }
            return counts;
        }

        /// <summary>Motes on standby, which are the currency for summons.</summary>
        public static ElementTable<int> StandbyCounts(IEnumerable<MoteInstance> motes, Func<string, MoteDef?> defOf)
        {
            var counts = ElementTable<int>.Filled(0);
            foreach (var mote in motes)
            {
                if (mote.State != MoteState.Standby) continue;
                var def = defOf(mote.DefId);
                if (def != null) counts[def.Element] += 1;
            }
            return counts;
        }

        /// <summary>Move a set mote to standby. False if it was not available.</summary>
        public static bool Unleash(IList<MoteInstance> motes, string defId)
        {
            var mote = motes.FirstOrDefault(m => m.DefId == defId && m.State == MoteState.Set);
            if (mote == null) return false;
            mote.State = MoteState.Standby;
            mote.RecoveryLeft = 0;
            return true;
        }

        /// <summary>
        /// Spend standby motes of an element on a summon, moving them to
        /// Recovering. Returns the ids spent, empty if the cost could not be
        /// met - in which case nothing is mutated.
        /// </summary>
        public static List<string> SpendStandby(IList<MoteInstance> motes, Element element, int count, Func<string, MoteDef?> defOf)
        {
            var available = motes
                .Where(m => m.State == MoteState.Standby && defOf(m.DefId)?.Element == element)
                .ToList();
            if (available.Count < count) return new List<string>();

            var spent = new List<string>();
            for (int i = 0; i < count; i++)
            {
                var mote = available[i];
                mote.State = MoteState.Recovering;
                mote.RecoveryLeft = defOf(mote.DefId)?.RecoveryRounds ?? DefaultRecoveryRounds;
                spent.Add(mote.DefId);
            }
            return spent;
        }

        /// <summary>End-of-round recovery tick. Returns ids that returned to Set.</summary>
        public static List<string> TickRecovery(IEnumerable<MoteInstance> motes)
        {
            var recovered = new List<string>();
            foreach (var mote in motes)
            {
                if (mote.State != MoteState.Recovering) continue;
                mote.RecoveryLeft -= 1;
                if (mote.RecoveryLeft <= 0)
                {
                    mote.State = MoteState.Set;
                    mote.RecoveryLeft = 0;
                    recovered.Add(mote.DefId);
                }
            }
            return recovered;
        }

        /// <summary>Between battles every mote returns to its bearer.</summary>
        public static void RestoreAll(IEnumerable<MoteInstance> motes)
        {
            foreach (var mote in motes)
            {
                mote.State = MoteState.Set;
                mote.RecoveryLeft = 0;
            }
        }
    }
}
